### PR TITLE
JVM_IR: refactor ClassCodegen and FunctionCodegen

### PR DESCRIPTION
**Pull request by pyos**
Non-behavior-affecting parts of #3056 and #3164. Depends on #3204.

### CODE REVIEW

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ClassCodegen.kt [hunk 1]
```diff
@@ -384,8 +353,26 @@ open class ClassCodegen protected constructor(
             return
         }
 
-        val signature = FunctionCodegen(method, this).generate().asmMethod
-        jvmSignatureClashDetector.trackMethod(method, RawSignature(signature.name, signature.descriptor, MemberKind.METHOD))
+        val node = FunctionCodegen(method, this).generate()
+        val mv = with(node) { visitor.newMethod(method.OtherOrigin, access, name, desc, signature, exceptions.toTypedArray()) }
```
Comment (max-kammerer) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ClassCodegen.kt]: NB: cached bytecode in general case would be totally different from writen one: cause optimizations are not perfromed (visitor.newMethod adds a bunch of optimization wrappers)

Comment (max-kammerer) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ClassCodegen.kt]: Oh, sorry it's not cached

**Comment by max-kammerer** [comment]
By unknown reason `org.jetbrains.kotlin.codegen.ir.FirBlackBoxCodegenTestGenerated.Coroutines.VarSpilling#testRefinedIntTypesAnalysis` test is unmuted. @pyos Please unmute

**Comment by pyos** [comment]
Done. The test no longer fails because the `suspend fun` in it is trivial so the only reason compilation fails is because it has no PSI element. With this refactoring, the check is not triggered unless a diagnostic on said element is actually needed.

**Comment by max-kammerer** [comment]
Thank you!

**Pull request merged**